analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotOverloadOperatorEqual.cs

```diff
@@ -29,27 +29,25 @@
 namespace SonarAnalyzer.Rules.CSharp
 {
     [DiagnosticAnalyzer(LanguageNames.CSharp)]
-    public sealed class DotNotOverloadOperatorEqual : SonarDiagnosticAnalyzer
+    public sealed class DoNotOverloadOperatorEqual : SonarDiagnosticAnalyzer
     {
-        internal const string DiagnosticId = "S3875";
+        private const string DiagnosticId = "S3875";
         private const string MessageFormat = "Remove this overload of 'operator =='.";
 
         private static readonly ImmutableArray<KnownType> InterfacesRelyingOnOperatorEqualOverload =
             ImmutableArray.Create(
                 KnownType.System_IComparable,
                 KnownType.System_IComparable_T,
-                KnownType.System_IEquatable_T
-            );
+                KnownType.System_IEquatable_T,
+                KnownType.System_Numerics_IEqualityOperators);
 
-        private static readonly DiagnosticDescriptor rule =
+        private static readonly DiagnosticDescriptor Rule =
             DescriptorFactory.Create(DiagnosticId, MessageFormat);
 
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
 
-        protected override void Initialize(SonarAnalysisContext context)
-        {
+        protected override void Initialize(SonarAnalysisContext context) =>
             context.RegisterSyntaxNodeActionInNonGenerated(CheckForIssue, SyntaxKind.OperatorDeclaration);
-        }
 
         private void CheckForIssue(SyntaxNodeAnalysisContext analysisContext)
         {
```

Comment:
Actually instead of a bunch of return statements, all the ifs could be merged and at the end inverted and you should end up with something like:
```
            var declaration = (OperatorDeclarationSyntax)analysisContext.Node;
            if (declaration.OperatorToken.IsKind(SyntaxKind.EqualsEqualsToken)
                && declaration.Parent is not StructDeclarationSyntax
                && declaration.Parent is ClassDeclarationSyntax classDeclaration
                && !classDeclaration.ChildNodes()
                       .OfType<OperatorDeclarationSyntax>()
                       .Any(op => op.OperatorToken.IsKind(SyntaxKind.PlusToken)
                                  || op.OperatorToken.IsKind(SyntaxKind.MinusToken))
                && analysisContext.SemanticModel.GetDeclaredSymbol(classDeclaration) is { } namedTypeSymbol
                && !namedTypeSymbol.ImplementsAny(InterfacesRelyingOnOperatorEqualOverload))
            {
                analysisContext.ReportIssue(Diagnostic.Create(Rule, declaration.OperatorToken.GetLocation()));
            }
```

Reply:
```csharp
declaration.Parent is not StructDeclarationSyntax
&& declaration.Parent is ClassDeclarationSyntax classDeclaration
```

Actually, do we even need both of these? Isn't the first redundant if you have the second one?